recipe-client-addon/lib/PreferenceManagement.jsm

```diff
@@ -0,0 +1,395 @@
+    store.saveSoon();
+  },
+
+
+  /**
+   * Returns the internal storage object, after properly ensuring the store and
+   * store namespace are initiated.
+   *
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async setStorageItem(key, value) {
+    const store = await this.getStorage();
+    store.data[namespace][key] = value;
+  },
+
+  /**
+   * Test wrapper that temporarily replaces the stored recipe data with fake
+   * data for testing. Given a test function, mocks the data store, then executes
+   * that function.
+   */
+  withMockData(testFunction) {
+    // We use a POJO to fake the internal storage.
+    this.mockData = {
+      [namespace]: {},
+    };
+
+    return async (...args) => {
+      this.storePromise = Promise.resolve({
+        data: this.mockData,
+        saveSoon() {},
+      });
+
+      try {
+        // Tests directly edit store values, so for test functions, the relevant
+        // namespace is exposed for brevity.
+        await testFunction(this.mockData[namespace], ...args)
+      } finally {
+        this.stopAllObservers();
+
+        // Clear up any remnants from the test function
+        this.clearStorage();
+
+        // Undo the mock storage
+        this.storePromise = undefined;
+      }
+
+      // ensureStorage will use the fake `storePromise`, returning mock data.
+      return await this.ensureStorage();
+    };
+  },
+
+  /**
+   * Looks up active recipes in the local store that modify preference values,
+   * and applies those changes across Default/User prefs.
+   *
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async applyAll() {
+    // On shutdown, clear all the observers that we're about to instantiate.
+    CleanupManager.addCleanupHandler(this.stopAllObservers.bind(this));
+
+    for (const recipe of await this.getActiveChanges()) {
+      // Select the User or Default branch, then update the recipe's preference
+      // with the appropriate value.
+      const prefs = this.getPreferenceBranch(recipe.preferenceBranchType);
+      prefs.set(recipe.preferenceName, recipe.preferenceValue);
+
+      // Notify Telemetry of recipes we're running, since they don't persist between restarts
+      TelemetryEnvironment.setExperimentActive(recipe.name, recipe.branch);
+
+      // Watch for changes to the recipe's preference
+      this.startObserver(recipe.name, recipe.preferenceName, recipe.preferenceValue);
+    }
+  },
+
+  /**
+   * Returns a reference to either the Default or User preferences branch
+   *
+   * @param  {string} which Branch to return ('user' or 'default')
+   * @return {Object}       Selected preferences branch
+   */
+  getPreferenceBranch(which) {
+    const branch = PreferenceBranchType[which];
+    if (!branch) {
+      throw new Error(`Invalid value for preference branch: "${which}"`);
```

Comment:
Nit: this would be slightly simpler as:

```js
if (!PreferenceBranchType.hasOwnProperty(which)) {
  throw new Error(...);
}
return PreferenceBranchType[which];
```